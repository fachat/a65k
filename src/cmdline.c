
/****************************************************************************

    commandline parsing
    Copyright (C) 2016-2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdlib.h>
#include <strings.h>
#include <string.h>

#include "hashmap.h"
#include "array_list.h"
#include "cmdline.h"
#include "err.h"
#include "infiles.h"

// hash from param name to cmdline_t sruct for quick find
static hash_t *params = NULL;
// list of parameters for help output
static list_t *paramlist = NULL;

static const char *prg_name = NULL;

static err_t usage(int flag, void* extra) {
	printf("Usage: %s [options] in-filename1 [in-filename2 ...]\n"
		"Cross-assembler for 65xx/R65C02/65816/65CE02/65002\n"
		"\n", prg_name); 

	list_iterator_t *iter = list_iterator(paramlist);
	cmdline_t *param = NULL;
	while ( (param = list_iterator_next(iter)) ) {
		printf("  -%s\n\t%s\n", param->name, param->description);
	}
	
	return E_ABORT;
}

static const cmdline_t help[] = {
	{ "?", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
	{ "help", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
};

static const char *key_from_param(const void *entry) {
	return ((cmdline_t*)entry)->name;
}

// TODO init the cmdline parser with the name as which the program was called
// for example "a65k", but also "xa65"
void cmdline_module_init() {

	params = hash_init_stringkey(50, 25, key_from_param);
	paramlist = array_list_init(20);

	cmdline_register_mult(help, sizeof(help)/sizeof(cmdline_t));
}

// template method where extra_param is the pointer to an int variable to set
err_t cmdline_set_flag(int flag, void *extra_param) {
	*((int*)extra_param) = flag;
	return E_OK;
}

void cmdline_register_mult(const cmdline_t *param, int num ) {
	for (int i = 0; i < num; i++) {
		cmdline_register(&param[i]);
	}
}

void cmdline_register(const cmdline_t *param) {
	
	if (hash_put(params, (void*)param) != NULL) {
		// must not happen
		exit(1);
	}
	list_add(paramlist, (void*)param);
}

err_t cmdline_parse(int argc, char *argv[]) {

	err_t rv = E_OK;

	prg_name = argv[0];

        int i = 1;
        while (i < argc && !rv) {
                if (argv[i][0] == '-') {
			char *val = NULL;
			int flag = 1;
			const char *name = argv[i]+1;
			if (name == strstr(name, "no-")) {
				name = name+3;
				flag = 0;
			}
			char *end = index(name, '=');
			if (end != NULL) {
				val = end + 1;
				end[0] = 0;
			}
			
			// lookup option
			cmdline_t *opt = hash_get(params, name);
			if (opt != NULL) {
				switch (opt->type) {
				case PARTYPE_FLAG:
					rv = opt->setflag(flag, opt->extra_param);
					break;
				case PARTYPE_PARAM:
					// TODO: error if param is missing
					// TODO: if started with "no-" return error
					rv = opt->setfunc(val, opt->extra_param);
					break;
				}
			} else {
				// TODO error cmdline parameter not found
			}
                } else {
			// register file with snapshot (clone) of current parser configuration
                        infiles_register(argv[i], parser_current_config());
                }   
		i++;
        }   
	return rv;
}




