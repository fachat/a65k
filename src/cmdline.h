
/****************************************************************************

    commandline parsing
    Copyright (C) 2016 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef CMDLINE_H
#define CMDLINE_H

#include "err.h"

// init the cmdline parser 
void cmdline_module_init();

// parse the options
err_t cmdline_parse(int argc, char *argv[]);

typedef enum {
	// no parameter, sets a bool when present with/without "no-" prefix
	PARTYPE_FLAG,
	// with required parameter
	PARTYPE_PARAM
} param_type_t;

typedef struct {
	// name of cmdline param
	const char 	*name;		

	// option has a parameter
	param_type_t	type;

	// function to call when parsed and type is PARTYPE_PARAM
	err_t 		(*setfunc)(const char *value, void *extra_param);

	// function to call when parsed and type is PARTYPE_FLAG
	void 		(*setflag)(int flag, void *extra_param);

	// extra param to be passed to set function
	void		*extra_param;

	// description
	const char	*description;
} cmdline_t;

// template method where extra_param is the pointer to an int variable to set
void cmdline_set_flag(int flag, void *extra_param);
 
void cmdline_register(const cmdline_t *param);

#endif

