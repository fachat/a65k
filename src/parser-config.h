/****************************************************************************

    parser
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PARSER_CONFIG_H
#define PARSER_CONFIG_H


typedef struct {
	// parse address and bytes (disables initial_lineno, default off)
	int	initial_binary;	

	// parse initial line number (default, disables initial_binary)
	int	initial_lineno;

	// do the actual include (default in assembler, off in lint mode)
	int	follow_includes;

	// parse xa preprocessor (default off)
	int	xa_preprocessor;

	// convert xa pp to standard pseudo-ops (only in lint mode)
	int	convert_xapp;

	// allow colons in comments (instead of starting a new statement)
	int	colon_in_comments;

	// allow C-style hex and octal
	int	cstyle_allowed;

} parser_config_t;

void parser_config_init(); 

// return a clone of the current configuration
parser_config_t *parser_current_config();

#endif

