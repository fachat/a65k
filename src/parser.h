/****************************************************************************

    parser
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PARSER_H
#define PARSER_H

#include "context.h"
#include "label.h"
#include "block.h"
#include "operation.h"
#include "tokenizer.h"
#include "arith.h"
#include "list.h"

typedef enum {
        S_LABEQPC,              // set label to PC
        S_LABDEF,               // set label from parameter
	S_OPCODE,		// opcode
	S_PSEUDO,		// pseudo opcode
} stype_t;

/**
 * The statement_t struct is the main structure of the internal workings.
 * The parser creates a statement_t with all information to be gathered from the source.
 * The printer can "lint" this statement and pretty-format it
 * The assembler itself resolves labels and computes references and extends it with binary values
 * The printer can then also print the extended statement with binary values
 * The renderer can then put (a list of) statements into a binary format, depending on the output options
 */
typedef struct {
        const block_t           *blk;		// pointer to current block structure for label dereferencing
        const context_t         *ctx;		// current context, like CPU type, width, other options
        stype_t                 type;		// type of statement (see above)
        // optional
        const label_t           *label;		// TODO allow for multiple label definitions per line/statement
	// S_LABDEF
	const ilist_t		*setlabel;	// anode_t list for setting label to a value 
	// S_OPCODE
       	const operation_t       *op;		// pointer to info about operation, like "ADC", "TAX", ...
	op_t			base;		// base extension of the parameter, like ",B", ",s" if any, zero else
	
	syntax_type		syn;		// addressing mode syntax, see SY_* constants
       	ilist_t  	        *param;		// list of anode_t structs that describe the syntax tree for the parameter value
} statement_t;


void parser_module_init(void);

err_t parser_push(const context_t *context, const line_t *line);

list_iterator_t *parser_get_statements(void);

// for test cases only
void parser_reset();

#endif

