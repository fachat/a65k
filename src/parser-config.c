/****************************************************************************

    parser
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <string.h>

#include "mem.h"
#include "err.h"
#include "parser-config.h"
#include "cmdline.h"


static type_t parser_config_memtype = {
        "parser_config_t",
        sizeof(parser_config_t)
};

static parser_config_t pconfig;

static err_t set_initial_binary(int flag, void *params) {
	(void) params;
	pconfig.initial_binary = flag;
	if (flag) {
		pconfig.initial_lineno = 0;
	}
	return E_OK;
}
static err_t set_initial_lineno(int flag, void *params) {
	(void) params;
	pconfig.initial_lineno = flag;
	if (flag) {
		pconfig.initial_binary = 0;
	}
	return E_OK;
}


static cmdline_t params[] = {
	{ "parse-initial-binary", PARTYPE_FLAG, NULL, set_initial_binary, NULL, 
		"parse address and hex data values before the actual label and operation "
		"(disables parse-initial-lineno)" },
	{ "parse-initial-lineno", PARTYPE_FLAG, NULL, set_initial_lineno, NULL, 
		"parse a BASIC-style line number before the actual label and operation " 
		"(disables parse-initial-binary)" },
	{ "parse-follow-includes", PARTYPE_FLAG, NULL, cmdline_set_flag, &pconfig.follow_includes, 
		"Follow include statements (default in assembler, off by default in lint mode)" },
	{ "parse-xa-preproc", PARTYPE_FLAG, NULL, cmdline_set_flag, &pconfig.xa_preprocessor, 
		"parse XA65 preprocessor statements (default off)" },
	{ "parse-convert-xapp", PARTYPE_FLAG, NULL, cmdline_set_flag, &pconfig.convert_xapp,
		"Convert XA65 preprocessor statements into standard pseudo-operations (lint mode only)" },
	{ "parse-colon-in-comments", PARTYPE_FLAG, NULL, cmdline_set_flag, &pconfig.colon_in_comments,
		"if set, allow colon in comments; by default colon starts new statement." },
	{ "parse-c-style", PARTYPE_FLAG, NULL, cmdline_set_flag, &pconfig.cstyle_allowed,
		"allow C-style hex and octal values (default off)" },
};

void parser_config_init() {

	pconfig.initial_binary = 0;
	pconfig.initial_lineno = 1;
	pconfig.follow_includes = 1;
	pconfig.xa_preprocessor = 0;
	pconfig.convert_xapp = 1;
	pconfig.colon_in_comments = 0;
	pconfig.cstyle_allowed = 0;	
	

	//printf("sizeof(params)=%ld\n", sizeof(params));

	int n = sizeof(params) / sizeof(cmdline_t);
	for (int i = 0; i < n; i++) {
		cmdline_register(&params[i]);
	}
}

parser_config_t* parser_current_config() {

	parser_config_t *clonedconf = mem_alloc(&parser_config_memtype);

	memcpy(clonedconf, &pconfig, sizeof(pconfig));

	return clonedconf;
}

