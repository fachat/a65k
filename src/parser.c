/****************************************************************************

    parser
    Copyright (C) 2015-2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#define	DEBUG

#include <stdio.h>
#include <ctype.h>

#include "err.h"

#include "log.h"
#include "mem.h"
#include "list.h"
#include "array_list.h"
#include "hashmap.h"
#include "position.h"

#include "cpu.h"
#include "segment.h"
#include "context.h"
#include "parser.h"
#include "label.h"
#include "block.h"
#include "tokenizer.h"
#include "arith.h"
#include "parser-config.h"

typedef struct {
	block_t			*blk;

	list_t			*statements;
} parser_t;


typedef enum {
	P_INIT,		// start of line, accept label definitions, operations and pseudos
	P_OP,		// has parsed label, accept "=" or operation
	P_PARAM,	// after operation, parse parameters to the operation
	P_COLON,	// after either label or line number, where a colon can be
	P_END,		// expect end of line
} pstate_t;


static const type_t statement_memtype = {
	"statement_t",
	sizeof(statement_t)
};

static const type_t parser_memtype = {
	"parser_t",
	sizeof(parser_t)
};

static parser_t *p = NULL;

void parser_module_init(void) {

	p = mem_alloc(&parser_memtype);
	p->blk = NULL;
	p->statements = array_list_init(10000);

	parser_config_init();
}


static statement_t *new_statement(const context_t *ctx) {
	statement_t *stmt = mem_alloc(&statement_memtype);

	stmt->type = S_NONE;
	stmt->blk = p->blk;
	stmt->ctx = ctx;
	stmt->op = NULL;
	stmt->label = NULL;
	stmt->syn = SY_IMP;

	// negative line number is not set
	stmt->lineno = -1;

	stmt->rs_prefix = RS_NOT_SET;
	stmt->le_prefix = LE_NOT_SET;
	stmt->um_prefix = 0;
	stmt->nf_prefix = 0;

	stmt->comment = NULL;

	return stmt;
}

static statement_t *new_statement_in_line(const context_t *ctx, statement_t *prev) {

	statement_t *stmt = new_statement(ctx);

	stmt->lineno = prev->lineno;

	return stmt;
}


static void statement_push(const context_t *ctx, statement_t *stmt) {

	// statements in parse order
        list_add(p->statements, stmt);

	// statements per segment/region
	segment_push_statement(ctx->segment, stmt);
}


void parser_reset() {
        // only for tests - ignore that we would have to mfree stuff before
        p->statements = array_list_init(10000);
}

static inline void error_double_prefix(const tokenizer_t *tok, char prefix, char conflict) {
        toklog_error(tok, "Double prefix '%c' conflicting with previous '%c'", prefix, conflict);
}

static inline void warn_double_prefix(const tokenizer_t *tok, char prefix) {
        toklog_warn(tok, "Double prefix '%c'", prefix);
}

static inline void warn_operation_not_for_cpu(const tokenizer_t *tok, const char *op_name, const char *cpu_name) {
        toklog_warn(tok, "Operation name %s is not available for CPU %s, assuming label!", op_name, cpu_name);
}

static inline void error_assign_without_label(const tokenizer_t *tok) {
	toklog_error(tok, "%s", "Assign without label");
}


/**
 * parse the operation parameter incl. addressing mode
 */
err_t parse_param(tokenizer_t *tok, statement_t *stmt) {

	if (tok->type == T_TOKEN && tok->vals.op == OP_HASH) {

		stmt->syn = SY_IMM;

		if (tokenizer_next(tok, 0)) {
			return arith_parse_integer(tok, stmt->blk, 0, (const ilist_t**)&stmt->param);	
		}
		return E_OK;
	} else {
		err_t rv = arith_parse_integer(tok, stmt->blk, 1, (const ilist_t**)&stmt->param);	
		if (rv == E_OK && stmt->param->len > 0) {

			stmt->syn = SY_ABS;

			anode_t *outer = NULL;
			anode_t *inner = NULL;
			anode_t *base = NULL;
			asub_type brtype = 0;
			anode_t *last = NULL;

			// unwrap brackets and/or index tokens (',x')

			// 1st unwrap ending indexes, just ",y", ",z"  
			last = ilist_last(stmt->param);
			if (last && last->type == A_INDEX && (last->op == OP_YIND || last->op == OP_ZIND)) {
				outer = ilist_pop(stmt->param);
			}

			// 2nd unwrap brackets
			anode_t *first = ilist_get(stmt->param, 0);
			if (first->type == A_BRACKET && ilist_len(stmt->param) == 1) {
				// outer node is a bracket, and it is the only one
				brtype = first->val.subv.type; 
				ilist_t *basenode = first->val.subv.value;
				stmt->param = basenode;

				anode_t *first = ilist_get(stmt->param, 0);
				if (brtype == AB_RND 
						&& first->type == A_BRACKET 
						&& first->val.subv.type == AB_RND 
						&& ilist_len(stmt->param) == 1) {
					brtype = AB_DBLRND;
					ilist_t *basenode = first->val.subv.value;
					stmt->param = basenode;
				}
			}

			// 3rd from resulting unwrap ,x index
			last = ilist_last(stmt->param);
			if (last && last->type == A_INDEX && last->op == OP_XIND) {
				inner = ilist_pop(stmt->param);
			}

			// 4th finally unwrapp base (,s / ,b / ,p) 
			last = ilist_last(stmt->param);
			if (last && last->type == A_INDEX) {
				base = ilist_pop(stmt->param);
				stmt->base = base->op;
			}

			// then compute syntax identifier
			// note that base is independent of syntax
			if (inner != NULL) {
				if (outer != NULL) {
					rv = E_SYNTAX;
				}
				switch (brtype) {
				case 0:
					stmt->syn = SY_ABSX;
					break;
				case AB_RND:
					stmt->syn = SY_XIND;
					break;
				case AB_RCT:
					stmt->syn = SY_XINDL;
					break;
				case AB_DBLRCT:
					stmt->syn = SY_XINDQ;
					break;
				case AB_DBLRND:
					stmt->syn = SY_XINDD;
					break;
				default:
					// borken
					rv = E_SYNTAX;
					break;
				}
			} else if (outer != NULL && outer->op == OP_YIND) {
				switch (brtype) {
				case 0:
					stmt->syn = SY_ABSY;
					break;
				case AB_RND:
					stmt->syn = SY_INDY;
					break;
				case AB_RCT:
					stmt->syn = SY_INDYL;
					break;
				case AB_DBLRCT:
					stmt->syn = SY_INDYQ;
					break;
				case AB_DBLRND:
					stmt->syn = SY_INDYD;
					break;
				default:
					// borken
					rv = E_SYNTAX;
					break;
				}
			} else if (outer != NULL && outer->op == OP_ZIND) {
				switch (brtype) {
				case 0:
					stmt->syn = SY_ABSZ;
					break;
				case AB_RND:
					stmt->syn = SY_INDZ;
					break;
				case AB_RCT:
					stmt->syn = SY_INDZL;
					break;
				case AB_DBLRCT:
					stmt->syn = SY_INDZQ;
					break;
				case AB_DBLRND:
					stmt->syn = SY_INDZD;
					break;
				default:
					// borken
					rv = E_SYNTAX;
					break;
				}
				
			
			}
		}
		return rv;
	}
}

static err_t parse_prefix(tokenizer_t *tok, statement_t *stmt) {

	err_t rv = E_OK;

	for (int i = 0; i < tok->len && !rv; i++) {
		char c = tolower(tok->line[tok->ptr + i]);
		switch(c) {
		case 'u':
			if (stmt->um_prefix) {
				warn_double_prefix(tok, c);
			}
			stmt->um_prefix = 1;
			break;
		case 'n':
			if (stmt->nf_prefix) {
				warn_double_prefix(tok, c);
			}
			stmt->nf_prefix = 1;
			break;
		case 'b':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_BYTE;
			break;
		case 'w':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_WORD;
			break;
		case 'l':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_LONG;
			break;
		case 'q':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_QUAD;
			break;
		case 'e':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_E;
			break;
		case 's':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_S;
			break;
		case '0':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_0;
			break;
		case '1':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_1;
			break;
		default:
			rv = E_SYNTAX;
			break;
		}
	}
	return rv;
}

err_t parser_push(const context_t *ctx, const line_t *line) {
	err_t rv = E_OK;

	if (line == NULL) {
		return E_END;
	}

	const parser_config_t *cfg = line->parsercfg;
	position_t *pos = line->position;

	// is the first block already set?
	if (p->blk == NULL) {
		p->blk = block_init(NULL, pos);
	}

	statement_t *stmt = new_statement(ctx);

	const instruction_t *op = NULL;
	const char *name = NULL;
	label_t *label = NULL;

	// tokenize the line
	tokenizer_t *tok = tokenizer_create(line->position, line->line, cfg->cstyle_allowed);

	tokenizer_next(tok, 0);

	while ((rv == E_OK) && tok->type != T_END && tok->is_valid) {

		while (tok->type == T_TOKEN && tok->vals.op == OP_COLON) {
			// ignore colon at beginning of line
			tokenizer_next(tok, 0);
		}

		// initial line number
		if (tok->type == T_LITERAL) {
			if (cfg->initial_lineno && (tok->vals.literal.type == LIT_DECIMAL)) {
				// line number parsing
				stmt->lineno = tok->vals.literal.value;

				tokenizer_next(tok, 0);
			} 

			if (tok->type == T_TOKEN && tok->vals.op == OP_COLON) {
				// ignore colon directly after line number
				tokenizer_next(tok, 0);
			}
		}

		while (stmt->op == NULL && tok->type == T_NAME) {
			name = mem_alloc_strn(tok->line + tok->ptr, tok->len);
			op = instruction_find(name);
			if (op != NULL) {
				// check if the operation is compatible with the current CPU
				if (0 == (ctx->cpu->isa & op->isa)) {
					// TODO: config for either no message or error
					warn_operation_not_for_cpu(tok, name, ctx->cpu->name);
					op = NULL;
				}
			}
			if (op == NULL) {
				// label
				// TODO: redefinition?
				label = label_init(ctx, name, pos);
				if (stmt->label) {
					// we already had a label
					stmt->type = S_LABEQPC;
					statement_push(ctx, stmt);
					stmt = new_statement_in_line(ctx, stmt);
				}
				stmt->label = label;
		
				tokenizer_next(tok, 0);
				if (tok->type == T_END) {
					stmt->type = S_LABEQPC;
				} else
				if (tok->type == T_TOKEN && tok->vals.op == OP_COLON) {
					// if we have a label with a colon, ignore colon
					tokenizer_next(tok, 0);
				}
			} else {
				// operation
				stmt->type = S_OPCODE;
				stmt->op = op;
				tokenizer_next(tok, 0);
				if (tok->type == T_TOKEN && tok->vals.op == OP_DOT) {
					// parse prefix
					if (tokenizer_next_prefix(tok) && (tok->type == T_NAME)) {
						rv = parse_prefix(tok, stmt);
					} else {
						error_syntax(tok, "Expecting prefix notation after '.'");
						rv = E_SYNTAX;
					}
					if (rv) {
						break;
					}
					tokenizer_next(tok, 0);
				}
	
				if (tok->type != T_TOKEN 
					|| (tok->vals.op != OP_SEMICOLON
						&& (tok->vals.op != OP_DOUBLESLASH || cfg->cstyle_allowed)
						&& (tok->vals.op != OP_COLON))) {
	
					// parse parameters
					rv = parse_param(tok, stmt);
	
					if (rv != E_OK) {
						break;
					}
				}
			}
		}

		if (!stmt->op) {
			if (tok->type == T_TOKEN && tok->vals.op == OP_DOT) {
				if (tokenizer_next(tok, 0)) {
					rv = parse_pseudo(tok, stmt);
					stmt->type = S_PSEUDO;
				} else {
					rv = E_SYNTAX;
				}
				if (rv) {
					break;
				}
			} else
			if (tok->type == T_TOKEN && tokenizer_op_details(tok->vals.op)->is_assign) {
				if (stmt->label == NULL && tok->vals.op != OP_ASSIGNMULT) {
					error_assign_without_label(tok);
					rv = E_SYNTAX;
					break;
				}
				if (stmt->label == NULL && tok->vals.op == OP_ASSIGNMULT) {
					stmt->type = S_PCDEF;
				} else {
					stmt->type = S_LABDEF;
				}
				stmt->assign = tok->vals.op;
				tokenizer_next(tok, 0);
				// parse parameters
				rv = parse_param(tok, stmt);
				if (rv) {
					break;
				}
			} else
			if (tok->type != T_END
				&& tok->vals.op != OP_SEMICOLON
				&& (tok->vals.op != OP_DOUBLESLASH || cfg->cstyle_allowed)
				&& (tok->vals.op != OP_COLON)) {
				rv = E_SYNTAX;
				error_syntax(tok, "Expecting opcode");
				break;
			}
		}

		if (tok->type == T_TOKEN) {
			if ((tok->vals.op == OP_SEMICOLON
				|| (tok->vals.op == OP_DOUBLESLASH && cfg->cstyle_allowed))) {

				if (tokenizer_next_comment(tok, cfg->colon_in_comments)) {
					stmt->comment = mem_alloc_strn(tok->line+tok->ptr, tok->len);

					tokenizer_next(tok, 0);
				}
			} 
		}

		if (tok->type == T_TOKEN) {
			if (tok->vals.op == OP_COLON) {
	
				statement_push(ctx, stmt);
				stmt = new_statement_in_line(ctx, stmt);

				tokenizer_next(tok, 0);
			} else {
				rv = E_SYNTAX;
				error_syntax(tok, "Expecting comment or colon");
			}
		}
	}

	if (rv == E_OK || rv == E_END) {
		statement_push(ctx, stmt);
	}
	tokenizer_free(tok);

	return rv;
}


list_iterator_t *parser_get_statements(void) {

	return list_iterator(p->statements);
}

