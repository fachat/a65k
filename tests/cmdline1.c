

#include "cmdline.h"
#include "parser.h"
#include "parser-config.h"


void print_cfg(parser_config_t *cfg) {

	printf("config:\n");
	printf("  initial_binary=%d\n", cfg->initial_binary);
}


int main(int argc, char *argv[]) {

	cmdline_module_init("a65k");
	parser_module_init();

	parser_config_t *cfg1 = parser_current_config();
	print_cfg(cfg1);

	char *pars[] = { "a65k", "-parse-initial-binary" };
	cmdline_parse(2, pars );

	parser_config_t *cfg2 = parser_current_config();
	print_cfg(cfg2);

	char *pars2[] = { "a65k", "-no-parse-initial-binary" };
	cmdline_parse(2, pars2 );

	parser_config_t *cfg3 = parser_current_config();
	print_cfg(cfg3);
}



