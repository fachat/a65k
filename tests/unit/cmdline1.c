

#include "cmdline.h"
#include "parser.h"
#include "parser-config.h"
#include "mem.h"

static err_t set_flag(int flag, void *extra) {
	printf("set_flag(%d)\n",flag);
	return E_OK;
}
static err_t set_inc(const char *val, void *extra) {
	printf("set_inc(%s)\n",val);
	return E_OK;
}
static err_t set_inc2(const char *val, void *extra) {
	printf("set_inc2(%s)\n",val);
	return E_OK;
}

cmdline_t par[] = {
        { "include-dir", "I", PARTYPE_PARAM, set_inc, NULL, NULL,
                "include dir", NULL },
        { NULL, "i", PARTYPE_PARAM, set_inc2, NULL, NULL,
                "include dir2", NULL },
        { NULL, "o", PARTYPE_FLAG, NULL, set_flag, NULL,
                "include flag", NULL },
};


void print_cfg(parser_config_t *cfg) {

	printf("config:\n");
	printf("  initial_binary=%d\n", cfg->initial_binary);
}


int main(int argc, char *argv[]) {

        (void)argc;
        (void)argv;

	cmdline_module_init("a65k");
	cmdline_register_mult(par, sizeof(par)/sizeof(cmdline_t));
	parser_module_init();

	parser_config_t *cfg1 = parser_current_config();
	print_cfg(cfg1);

	char *pars[] = { "a65k", "--parse-initial-binary" };
	cmdline_parse(2, pars );

	parser_config_t *cfg2 = parser_current_config();
	print_cfg(cfg2);

	char *pars2[] = { "a65k", "--no-parse-initial-binary" };
	cmdline_parse(2, pars2 );

	parser_config_t *cfg3 = parser_current_config();
	print_cfg(cfg3);

	char *pars3[] = { "a65k", mem_alloc_str("--include-dir=/foo"), mem_alloc_str("-I/bar") };
	cmdline_parse(3, pars3 );

	char *pars5[] = { "a65k", mem_alloc_str("-oifub") };
	cmdline_parse(2, pars5 );

	char *pars6[] = { "a65k", mem_alloc_str("+oiyyy") };
	cmdline_parse(2, pars6 );

	char *parsx[] = { "a65k", mem_alloc_str("-?") };
	cmdline_parse(2, parsx );
}



