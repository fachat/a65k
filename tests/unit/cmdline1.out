config:
  initial_binary=0
config:
  initial_binary=1
config:
  initial_binary=0
set_inc(/foo)
set_inc(/bar)
set_flag(1)
set_inc2(fub)
set_flag(0)
set_inc2(yyy)
Usage: a65k [options] in-filename1 [in-filename2 ...]
Cross-assembler for 65xx/R65C02/65816/65CE02/65002
Flag options can be inverted with 'no-' prefix for long
names resp. '+' for short names

  --help
  -?
	Show this help
  --include-dir=<value>
  -I<value>
	include dir
  -i<value>
	include dir2
  -o
	include flag
  --parse-initial-binary
	parse address and hex data values before the actual label and operation (disables parse-initial-lineno)
  --parse-initial-lineno
	parse a BASIC-style line number before the actual label and operation (disables parse-initial-binary)
  --parse-c-style
	allow C-style hex and octal values (default off)
unknown cmdline parameter: -?
